### ISSUE TITLE
Missing source maps break client-side debugging under VSCode

### ISSUE THREAD

**Issue by carlo-salinari** [vercel/next.js#56702]
### Link to the code that reproduces this issue

https://github.com/carlo-salinari/debug-nextjs

### To Reproduce

1. Start the application with `npm run dev`
2. In vscode, set a breakpoint at line 6 of `page.tsx`
3. Start a debugging session
4. Refresh page

### Current vs. Expected behavior

Current behaviour:
- breakpoint goes gray (ignored)
- debug console say some source maps file are missing (see attachment)
[dbconsole.txt](https://github.com/vercel/next.js/files/12870120/dbconsole.txt)

Expected:
- hit the breakpoint and let you debug the app

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #34~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Sep  7 13:12:03 UTC 2
Binaries:
  Node: 19.4.0
  npm: 9.2.0
  Yarn: 1.22.19
  pnpm: 8.9.0
Relevant Packages:
  next: 13.5.5-canary.5
  eslint-config-next: 13.5.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

**Comment by Rishab49** [comment]
@carlo-salinari try replacing the content of `.vscode/launch.json` with following

```
{
    "version": "0.2.0",
    "configurations": [
      {
        "name": "Next.js: debug server-side",
        "type": "node-terminal",
        "request": "launch",
        "command": "pnpm run dev"
      },
      {
        "name": "Next.js: debug client-side",
        "type": "pwa-chrome",
        "request": "launch",
        "url": "http://localhost:3000"
      },
      {
        "name": "Next.js: debug full stack",
        "type": "node-terminal",
        "request": "launch",
        "command": "pnpm run dev",
        "console": "integratedTerminal",
        "serverReadyAction": {
          "pattern": "started server on .+, url: (https?://.+)",
          "uriFormat": "%s",
          "action": "debugWithChrome"
        }
      }
    ]
  }


```

then, close all dev server and start debugging session in full stack mode

**Comment by carlo-salinari** [comment]
This was helpful, thanks.

Your `launch.json` has some invalid properties (`"type": "pwa-chrome"` and `"console": "integratedTerminal"`), but pointed me to the right direction.

Here's the one I ended up using (I've updated the reproduction repo):

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Attach Chrome",
      "type": "chrome",
      "request": "attach",
      "port": 9222,
      "url": "http://localhost:3000"
    },
    {
      "name": "Next.js: debug full stack",
      "type": "node-terminal",
      "request": "launch",
      "command": "npm run dev",
      "serverReadyAction": {
        "pattern": "- Local:.+(https?://.+)",
        "uriFormat": "%s",
        "action": "debugWithChrome"
      }
    }
  ]
}

```

I adapted it from [the official documentation](https://nextjs.org/docs/pages/building-your-application/configuring/debugging) (it's in the Pages Router section, there's no Debugging section in the App Router section).

`"Next.js: debug full stack"` works (but opens a new browser every time).
Attaching to a running `"use client"` page in chrome doesn't seem to work (breakpoint is not hit).

At this point I don't even know if the problem is with nextjs, vscode or chrome.

**Comment by Rishab49** [comment]
@carlo-salinari maybe `use client` is not debuggable because it gets removed at compile time.

**Comment by jelling** [comment]
> @carlo-salinari maybe `use client` is not debuggable because it gets removed at compile time.

At least in my case, that was the issue. Component A included Component B which used `use client.` As a result, Component A was no longer server-side debuggable. 

I'm relatively new to Next.Js so I'm not certain if that is working as intended and it's just a matter of communicating this better. Or if this is a bug.

**Comment by heecheon92** [comment]
Mine got fixed by changing the pattern for serverReadyAction in launch.json

from

```json
        "serverReadyAction": {
          "pattern": "started server on .+, url: (https?://.+)",
          "uriFormat": "%s",
          "action": "debugWithChrome"
        }
```

to

```json
      "serverReadyAction": {
        "pattern": "- Local:.+(https?://.+)",
        "uriFormat": "%s",
        "action": "debugWithChrome"
      }
```

**Comment by prurph** [comment]
Has anyone figured out how to debug server-side by attaching to an already running process, instead of launching with `node-terminal`?

My issue is if I use a config like:

```json
    {
      "type": "pwa-node",
      "name": "attach test",
      "request": "attach",
      "processId": "${command:PickProcess}",
      "cwd": "${workspaceFolder}"
    },
```

and then start NextJS with `next dev` (with or without `NODE_OPTIONS='--inspect'), the debugger can attach, but breakpoints do not fire.

VSCode will say "breakpoint can't be set" and offers debugging help that seems to suggest source maps aren't being generated. Indeed, if I attach using `chrome://inspect` and hit `CMD+P`, node of my source files are listed, only node internals and `node_module` stuff. The docs say source maps are generated automatically in dev, so I'm not sure what is going on.

**Comment by Skarvion** [comment]
Seems that the discussion derailed from debugging client-side to server-side. I just fixed my client-side debugging issue today, and I find that the debugging documentation for TypeScript is lacking.

# Solution

@carlo-salinari , I assume your issue is similar to mine, which is the problem on how to set up the `sourceMapPathOverrides` and `sourceMaps`.

The setup is dependent whether you use Turbopack or not. 

## Without Turbopack

Credit to [this article](https://thundermiracle.com/blog/en/2020-12-23-debug-nextjs-typescript-in-vscode/) although there's a typo with `webRoot` since it includes `src` in the path again. I've tweaked my code below to drop that.

```json
{
  "name": "Next.js: debug client-side",
  "type": "msedge",
  "request": "launch",
  "url": "http://localhost:3000",
  "webRoot": "${workspaceFolder}",
  "sourceMaps": true,
  "sourceMapPathOverrides": {
    "webpack://_N_E/*": "${webRoot}/*"
  }
}
```

## With Turbopack

```json
{
  "name": "Next.js: debug client-side",
  "type": "msedge",
  "request": "launch",
  "url": "http://localhost:3000",
  "webRoot": "${workspaceFolder}",
  "sourceMaps": true,
  "sourceMapPathOverrides": {
    "/turbopack/[project]/*": "${webRoot}/*"
  }
}
```

# How to diagnose the problem in the future

TLDR; use VSCode debug diagnostic to match the sourcemap path

I fixed the above problem by opening the devtool in the browser first and see the source tree first, to get a better feel of the TypeScript file structure

![image](https://github.com/vercel/next.js/assets/17715991/00719e79-3636-41ba-bf85-ebb1e7865232)

Then in VSCode, you open Debug Diagnostics (accessible via Command Palette). Select "What scripts and sourcemaps are loaded". Filter for the TSX file you're looking for, then you can grab the referenced from sourcemap to your `launch.json`

<img width="579" alt="image" src="https://github.com/vercel/next.js/assets/17715991/389b8d6a-2ab9-4a9e-8871-47c25d141993">

**Comment by carlo-salinari** [comment]
@Skarvion thanks for the hint.
I don't use Turbopack yet.
The Debug Diagnostics tool looks useful, but it is nowhere to be found in my VSCcode Command Palette (v1.85.2).
Are you sure it is not from an Extension? Or maybe there is some intermediate step to activate it?

**Comment by Skarvion** [comment]
@carlo-salinari I just checked my environment, it comes from an extension called [JavaScript Debugger](https://marketplace.visualstudio.com/items?itemName=ms-vscode.js-debug). Apparently it's a built-n extension, it should already be enabled for VSCode 1.46 onward. First time seeing a built-in extension, you can't even check it with `@installed`

Anyway, step by step:
1. Start Debugging (F5)
2. When you're in debug mode, open Debug Diagnostic through either:

-   Open Command Palette and look for "Debug: Diagnose Breakpoint Problems". Only available when you're in debug mode
-   Put a breakpoint somewhere and if it's greyed out, then if you hover your mouse, it offers more information and a button to troubleshoot

![image](https://github.com/vercel/next.js/assets/17715991/824004e2-24c2-4834-b841-832b5df2a163)

**Comment by Stan-Stani** [comment]
Fyi, my solution, where the client-side breakpoints just stopped working was to close (and reopen) all my VSCode instances and then retry.

**Comment by porkopek** [comment]
Thank you very much, @carlo-salinari .

Now I got turbopack working with debugging, the lack of which prevented me to use it

For server and client side debugging, I just added
```json
    { "name": "turbo",
      "type": "node-terminal",
      "request": "launch",
      "command": "npm run turbo",
      "serverReadyAction": {
        "pattern": "- Local:.+(https?://.+)",
        "uriFormat": "%s",
        "action": "debugWithChrome"
      },
      "sourceMapPathOverrides": {
        "/turbopack/[project]/*": "${webRoot}/*"
      }
    }
```
    
    where `npm run turbo` is `  an script like: "turbo": "next dev --turbo",`

**Comment by andrewmartin** [comment]
Hi everyone, thanks for your help and digging into this––

Curious if y'all have had any luck with getting server side files to work, cross-posting from https://github.com/vercel/next.js/issues/62008#issuecomment-1944338999

**Comment by DrPye** [comment]
@andrewmartin 
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Next.js: debug server-side",
      "type": "node-terminal",
      "request": "launch",
      "command": "npm run dev --turbo"
    },
    {
      "name": "Next.js: debug client-side",
      "type": "chrome",
      "request": "launch",
      "url": "http://localhost:3000"
    },
    {
      "name": "Next.js: debug full stack",
      "type": "node-terminal",
      "request": "launch",
      "command": "npm run dev --turbo",
      "skipFiles": ["<node_internals>/**"],
      "sourceMaps": true,
      "sourceMapPathOverrides": {
        "/turbopack/[project]/*": "${webRoot}/*"
      },
      "serverReadyAction": {
        "action": "debugWithEdge",
        "killOnServerStop": true,
        "pattern": "- Local:.+(https?://.+)",
        "uriFormat": "%s",
        "webRoot": "${workspaceFolder}"
      }
    }
  ]
}
```

**Comment by MiguelG97** [comment]
In my case, I was not able to debug with vscode due to source map issues when a the new vscode version 1.95 came up.
I fixed the problem by downgrading the vscode version. It looks that the problem is also related with the microsoft team because with the downgrade I can debug using the following launch json setup:

```
  "configurations": [
    {
      "type": "chrome",
      "request": "launch",
      "name": "Launch Chrome against localhost",
      "url": "http://localhost:8080",
      "webRoot": "${workspaceFolder}"
    }
  ]
```

**Comment by niko-la-petrovic** [comment]
Hey guys. Here's a working example for VSC on Windows.

Fresh Next.js v15.2.3 app created with `npx create-next-app@latest client-debugging` and turbopack enabled.

launch.json
```
{
   {
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Next.js: debug client-side",
      "type": "chrome",
      "request": "launch",
      "url": "http://localhost:3000",
      "webRoot": "${workspaceFolder}",
      "sourceMaps": true,
      "sourceMapPathOverrides": {
        "file://C%3A/Users/nikol/repos/nikola-petrovic/nextjs-tests/client-debugging/*": "${workspaceFolder}/*"
      }
    }
  ]
}
```

As the other guys suggested, hovering over the breakpoint reveals VSC couldn't find the corresponding source file. Troublshooting the breakpoint reveals the following

![Image](https://github.com/user-attachments/assets/10ea456a-5039-4c71-808e-94c46563c97a)

So in line with that, I wrote a `sourceMapPathOverrides` entry and it started working.

I'm not familiar with the internals at play here, but I find it very weird that this isn't made to work out of the box without needing any overrides 🤨

**Comment by gethari** [comment]
Anyone have a solution for monorepo with turborepo + PNPM ?

**Comment by mtranter** [comment]
I just stumbled on this solution and it seems to be working for Turborepo monorepo + pnpm. Fingers crossed!!

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Next.js: Debug full stack",
      "type": "node",
      "request": "launch",
      "program": "${workspaceFolder}/apps/web/node_modules/next/dist/bin/next",
      "args": ["--port", "3001"],
      "runtimeArgs": ["--inspect"],
      "skipFiles": ["<node_internals>/**"],
      "cwd": "${workspaceFolder}/apps/web",
      "serverReadyAction": {
        "action": "startDebugging",
        "killOnServerStop": true,
        "pattern": "- Local:.+(https?://.+)",
        "config": {
          "type": "chrome",
          "request": "launch",
          "url": "http://localhost:3001",
          "cwd": "${workspaceFolder}/apps/web",
          "port": 9222
        }
      }
    }
  ]
}
```

**Comment by marcoshernanz** [comment]
Thanks for the report. I retested the reproduction with a current VS Code debug adapter and the client breakpoint binds and pauses correctly. The missing `runtime.js.map` and `internal/helpers.js.map` files from the original log were added in #66254 and have shipped since Next.js 15. The remaining React Refresh TypeScript source warning is tracked in #80531, so I’m closing this as fixed.
